contracts/ERC20.sol

```diff
@@ -1,34 +1,112 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.0;
 
-import "./IERC20.sol";
+import "./IERC20Metadata.sol";
 
 contract ERC20 is IERC20 {
-  function totalSupply() external view returns (uint256) {
-    return 0;
+  uint256 public constant MAXINT = 2**256 - 1;
+  uint256 private _totalSupply;
+  mapping(address => uint256) private _balances;
+  string private _coinName;
+  string private _coinSymbol;
+  uint8 private _coinDecimals;
+  mapping(address => mapping(address => uint256)) private _allowances;
+
+  constructor(uint256 _initialSupply, string memory _name, string memory _symbol, uint8 _decimals) public {
+    _coinName = _name;
+    _coinSymbol = _symbol;
+    _coinDecimals = _decimals;
+    mint(msg.sender, _initialSupply);
+  }
+
+  function name() external view returns (string memory) {
+    return _coinName;
+  }
+  function symbol() external view returns (string memory) {
+    return _coinSymbol;
+  }
+  function decimals() external view returns (uint8) {
+    return _coinDecimals;
+  }
+
+  function totalSupply() external view override returns (uint256) {
+    return _totalSupply;
   }
 
-  function balanceOf(address account) external view returns (uint256) {
-    return 0;
+  function balanceOf(address account) external view override returns (uint256) {
+    return _balances[account];
   }
 
-  function transfer(address to, uint256 amount) external returns (bool) {
+  function transfer(address to, uint256 amount) external override returns (bool) {
+    // Check if the amount is greater than the balance of the sender
+    require(_balances[msg.sender] >= amount, "Not enough balance");
+    // Check if the recipient is not the zero address
+    require(to != address(0), "Invalid recipient");
+    // Decrease the sender's balance
+    _balances[msg.sender] -= amount;
+    // Increase the recipient's balance
```

Comment:
You have a bit too many comments 🤣 Good code is when there is no need to comment, or only comments for high-level ideas